File: app/src/main/java/com/osrsd/cache/provider/SpotAnimationProvider.java
package com.osrsd.cache.provider;

import com.osrsd.cache.def.SpotAnimationDefinition;
import com.osrsd.spring.domain.Definition;
import lombok.extern.slf4j.Slf4j;

import java.nio.ByteBuffer;
import java.util.stream.IntStream;

@Slf4j
public final class SpotAnimationProvider {

    public static Definition decode(ByteBuffer buffer, SpotAnimationDefinition definition) {
        do {
            int opcode = buffer.get() & 0xff;
            if (opcode == 0) {
                break;
            }

            if (opcode == 1) {
                definition.setModelId(buffer.getShort() & 0xffff);
            } else if (opcode == 2) {
                definition.setAnimationId(buffer.getShort() & 0xffff);
            } else if (opcode == 4) {
                definition.setResizeX(buffer.getShort() & 0xffff);
            } else if (opcode == 5) {
                definition.setResizeY(buffer.getShort() & 0xffff);
            } else if (opcode == 6) {
                definition.setRotation(buffer.getShort() & 0xffff);
            } else if (opcode == 7) {
                definition.setAmbient(buffer.get() & 0xff);
            } else if (opcode == 8) {
                definition.setContrast(buffer.get() & 0xff);
            } else if (opcode == 40) {
                int size = buffer.get() & 0xff;
                definition.setRecolorToFind(new short[size]);
                definition.setRecolorToReplace(new short[size]);
                IntStream.range(0, size).forEach(var4 -> {
                    definition.getRecolorToFind()[var4] = (short) (buffer.getShort() & 0xffff);
                    definition.getRecolorToReplace()[var4] = (short) (buffer.getShort() & 0xffff);
                });
            } else if (opcode == 41) {
                int size = buffer.get() & 0xff;
                definition.setTextureToFind(new short[size]);
                definition.setTextureToReplace(new short[size]);
                IntStream.range(0, size).forEach(var4 -> {
                    definition.getTextureToFind()[var4] = (short) (buffer.getShort() & 0xffff);
                    definition.getTextureToReplace()[var4] = (short) (buffer.getShort() & 0xffff);
                });
            } else {
                log.warn(String.format("Unhandled definition opcode with id: %s.", opcode));
            }
        } while (true);
        return definition;
    }

}

File: app/src/main/kotlin/com/osedit/cache/provider/SpotAnimationProvider.kt
package com.osedit.cache.provider

import com.osedit.spring.domain.Definition
import com.osedit.spring.domain.SpotAnimationDefinition
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer

object SpotAnimationProvider : Provider<SpotAnimationDefinition> {

    private val log: Logger = LoggerFactory.getLogger(SpotAnimationProvider::class.java)

    override fun decode(buffer: ByteBuffer, definition: SpotAnimationDefinition): Definition {
        do {
            when (val opcode: Int = buffer.get().toInt() and 0xff) {
                1 -> definition.modelId = buffer.short.toInt() and 0xffff
                2 -> definition.animationId = buffer.short.toInt() and 0xffff
                4 -> definition.resizeX = buffer.short.toInt() and 0xffff
                5 -> definition.resizeY = buffer.short.toInt() and 0xffff
                6 -> definition.rotation = buffer.short.toInt() and 0xffff
                7 -> definition.ambient = buffer.get().toInt() and 0xff
                8 -> definition.contrast = buffer.get().toInt() and 0xff
                40 -> {
                    val length: Int = buffer.get().toInt() and 0xff
                    definition.recolorToFind = ShortArray(length)
                    definition.recolorToReplace = ShortArray(length)
                    (0 until length).forEach {
                        definition.recolorToFind!![it] = (buffer.short.toInt() and 0xffff).toShort()
                        definition.recolorToReplace!![it] = (buffer.short.toInt() and 0xffff).toShort()
                    }
                }
                41 -> {
                    val length: Int = buffer.get().toInt() and 0xff
                    definition.textureToFind = ShortArray(length)
                    definition.textureToReplace = ShortArray(length)
                    (0 until length).forEach {
                        definition.textureToFind!![it] = (buffer.short.toInt() and 0xffff).toShort()
                        definition.textureToReplace!![it] = (buffer.short.toInt() and 0xffff).toShort()
                    }
                }
                0 -> break
                else -> log.warn(String.format("Unhandled definition opcode with id: %s.", opcode))
            }
        } while (true)
        return definition
    }

}
